### DISCUSSION TITLE
psycopg2 and async/sync context

### DISCUSSION THREAD

**Discussion by micdeved** [psycopg/psycopg2#1596] [Q&A]
This is not a question, this is not a bug report. I'm not sure what is this. It is for history and maybe it will be helpful for someone else.

It is known and recommended to use:
- psycopg2 - for production;
- psycopg2-binary for local development.

But in case when we use **psycopg2** on production we often get following error:
`You cannot call this from an async context - use a thread or sync_to_async.`

Which is caused by the `async_unsafe` attribute and `SynchronousOnlyOperation` error.

As a workaround we can use next env var `DJANGO_ALLOW_ASYNC_UNSAFE=1`

There is no trouble just using `psycopg2-binary` package.

Versions:
- Alpine Linux: 3.18.*;
- Python: 3.11.4;
- Django: 3.2.* (wsgi app);
- psycopg2, psycopg2-binary: 2.9.6.

P.S.: we do not use explicitly `async` code in a project.

**Comment by dvarrazzo** [comment]
psycopg2-binary is just a precompiled version of psycopg. Everything else
you mention are Django features: you should contact th for your enquiry.

**Reply by micdeved** [reply]
i understand it, but it is definitely strange if i replace `psycopg2` with `psycopg2-binary` then everything else is just fine without any additional changes.
